[query] Add `hl.query_table` function for reading tables with indices…

tpoterba opened the pull request:
… from Python

tpoterba commented:
Important infrastructure for Seqr.

patrick-schultz commented on hail/python/hail/ir/ir.py:
```diff
@@ -2865,6 +2866,65 @@ def renderable_bindings(self, i, default_value=None):
+
+    def _handle_randomness(self, create_uids):
+        if create_uids:
+            self.row_typ_with_uid = tstruct(**self.row_typ, __uid=ttuple(tint64, tint64))
+            self.drop_uid = False
+            self.reader.set_uid_field('__uid')
```
All the other `_handle_randomness` methods return new nodes, they don't modify the node. I think that's necessary, since an IR can be reused in different contexts.

patrick-schultz commented on hail/python/test/hail/table/test_table.py:
```diff
@@ -2016,3 +2016,60 @@ def test_indexed_read_boundaries(branching_factor):
+
+
+def test_query_table():
+    f = new_temp_file(extension='ht')
```
Maybe include a few tests with a compound key?

patrick-schultz commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -712,6 +717,245 @@ case class PartitionNativeReader(spec: AbstractTypedCodecSpec, uidFieldName: Str
+            ((cb, elt) => elt),
+            bound = "lower",
+            compareEnd(_, _, _, _ < 1)
+          ).search(cb, partitionerRuntime, EmitCode.present(cb.emb, endAndSignTuple)))
```
From a quick look, I think you've reimplemented the registered function `partitionerFindIntervalRange`. If so, you should just factor out the implementation to a staged function (as most of the interval registered functions are) and call that. I think we should try to keep all the staged interval ordering stuff in one place, since it's relatively complicated.

tpoterba commented:
good comments; addressed these and did a fair bit of redesigning of the Python stuff. Refactored the scala to use that method you mentioned, also got mad that we didn't already have a SStackInterval and added that.
